From 78199aa0bc91962fdadb55f3c5a5224f08606be4 Mon Sep 17 00:00:00 2001
From: Alison Chaiken <alison_chaiken@mentor.com>
Date: Wed, 8 Jul 2015 16:46:47 +0200
Subject: [PATCH] gstomx: find buffer private data in FTB and ETB methods

Add  definitions of struct GstOmxPlatformPrivateInfo member values
and print them for debugging.

Change-Id: Idf7eec60250b7e219cc81875b2c3fe3ed95d1417
Signed-off-by: Bharathraj Nagaraju <snbraj@codeaurora.org>
---
 omx/gstomx.c         | 30 +++++++++++++++++++++++
 omx/gstomx_meta.h    | 67 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 omx/gstomxvideodec.c | 65 ++++++++++++++++++++++++++++++++++++++++++++++----
 3 files changed, 158 insertions(+), 4 deletions(-)
 create mode 100644 omx/gstomx_meta.h

diff --git a/omx/gstomx.c b/omx/gstomx.c
index 9a77e5b..d1b83fe 100644
--- a/omx/gstomx.c
+++ b/omx/gstomx.c
@@ -45,6 +45,12 @@
 #include "gstomxanalogaudiosink.h"
 #include "gstomxhdmiaudiosink.h"
 
+#define USE_OMX_TARGET_VENUS
+
+#if defined (USE_OMX_TARGET_VENUS)
+#include "gstomx_meta.h"
+#endif
+
 GST_DEBUG_CATEGORY (gstomx_debug);
 #define GST_CAT_DEFAULT gstomx_debug
 
@@ -580,6 +586,20 @@ EventHandler (OMX_HANDLETYPE hComponent, OMX_PTR pAppData, OMX_EVENTTYPE eEvent,
   return OMX_ErrorNone;
 }
 
+static void findMeta (OMX_BUFFERHEADERTYPE *bufhead) {
+
+  GstOmxPlatformPrivateInfo *ionBufInfo;
+  if ((GstOmxPlatformPrivateInfo *)((GstOmxPlatformPrivateList *)bufhead->pPlatformPrivate))
+    ionBufInfo = (GstOmxPlatformPrivateInfo *)((GstOmxPlatformPrivateList *)bufhead->pPlatformPrivate)->entryList->entry;
+  else {
+    g_print("buffer at %p has AllocLen %u and FilledLen %u\n", bufhead->pBuffer,(unsigned int) bufhead->nAllocLen, (unsigned int) bufhead->nFilledLen);
+    return;
+  }
+
+  g_print("buffer at %p has AllocLen %u, FilledLen %u, fd %u, offset %u  and ion_size %u\n", bufhead->pBuffer, (unsigned int) bufhead->nAllocLen, (unsigned int) bufhead->nFilledLen, ionBufInfo->ion_buf_fd, ionBufInfo->ion_buf_offset, (unsigned int) ionBufInfo->ion_buf_size);
+  return;
+}
+
 static OMX_ERRORTYPE
 EmptyBufferDone (OMX_HANDLETYPE hComponent, OMX_PTR pAppData,
     OMX_BUFFERHEADERTYPE * pBuffer)
@@ -588,6 +608,11 @@ EmptyBufferDone (OMX_HANDLETYPE hComponent, OMX_PTR pAppData,
   GstOMXComponent *comp;
   GstOMXMessage *msg;
 
+  /* debug dmabuf = ionBuf
+    g_print("In EmptyBufferDone ");
+    findMeta(pBuffer);
+  */
+
   buf = pBuffer->pAppPrivate;
   if (!buf) {
     GST_ERROR ("Have unknown or deallocated buffer %p", pBuffer);
@@ -626,6 +651,11 @@ FillBufferDone (OMX_HANDLETYPE hComponent, OMX_PTR pAppData,
   GstOMXComponent *comp;
   GstOMXMessage *msg;
 
+  /* debug dmabuf = ionBuf
+     g_print("In FillBufferDone ");
+     findMeta(pBuffer);
+  */
+
   buf = pBuffer->pAppPrivate;
   if (!buf) {
     GST_ERROR ("Have unknown or deallocated buffer %p", pBuffer);
diff --git a/omx/gstomx_meta.h b/omx/gstomx_meta.h
new file mode 100644
index 0000000..b371e68
--- /dev/null
+++ b/omx/gstomx_meta.h
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2016, The Linux Foundation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ *       copyright notice, this list of conditions and the following
+ *       disclaimer in the documentation and/or other materials provided
+ *       with the distribution.
+ *     * Neither the name of The Linux Foundation nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef __GST_OMX_META_H__
+#define __GST_OMX_META_H__
+
+G_BEGIN_DECLS
+
+const char *decode_color(OMX_COLOR_FORMATTYPE colorcode);
+
+typedef struct GstOmxPlatformPrivateInfo {
+  /** ion buffer file descriptor */
+  unsigned int ion_buf_fd;
+  /** Offset from ion buffer base address */
+  unsigned int ion_buf_offset;
+  /** Size of ion buffer */
+  unsigned int ion_buf_size;
+} GstOmxPlatformPrivateInfo;
+
+typedef struct GstOmxPlatformPrivateEntry {
+  /** Entry type */
+  int type;
+
+  /** Pointer to platform specific entry */
+  void *entry;
+} GstOmxPlatformPrivateEntry;
+
+typedef struct GstOmxPlatformPrivateList {
+  /** Number of entries */
+  unsigned int nEntries;
+
+  /** Pointer to array of platform specific entries *
+   * Contiguous block of GstOmxPlatformPrivateEntry element
+  */
+  GstOmxPlatformPrivateEntry *entryList;
+} GstOmxPlatformPrivateList;
+
+G_END_DECLS
+
+#endif /* __GST_OMX_META_H__ */
diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 8590301..22581c7 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -36,6 +36,8 @@
 #pragma GCC optimize ("gnu89-inline")
 #endif
 
+#define USE_OMX_TARGET_VENUS
+
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
 #include <gst/gl/gl.h>
 #include <gst/gl/egl/gsteglimagememory.h>
@@ -51,6 +53,9 @@
 #include "gstomxbufferpool.h"
 #include "gstomxvideo.h"
 #include "gstomxvideodec.h"
+#if defined (USE_OMX_TARGET_VENUS)
+#include "gstomx_meta.h"
+#endif
 
 GST_DEBUG_CATEGORY_STATIC (gst_omx_video_dec_debug_category);
 #define GST_CAT_DEFAULT gst_omx_video_dec_debug_category
@@ -1711,6 +1716,57 @@ gst_omx_video_dec_stop (GstVideoDecoder * decoder)
   return TRUE;
 }
 
+const char *decode_color(OMX_COLOR_FORMATTYPE colorcode) {
+  switch(colorcode) {
+  case OMX_COLOR_FormatUnused: return "OMX_COLOR_FormatUnused";
+  case OMX_COLOR_FormatMonochrome: return "OMX_COLOR_FormatMonochrome";
+  case OMX_COLOR_Format8bitRGB332: return "OMX_COLOR_Format8bitRGB332";
+  case OMX_COLOR_Format12bitRGB444: return "OMX_COLOR_Format12bitRGB444";
+  case OMX_COLOR_Format16bitARGB4444: return "OMX_COLOR_Format16bitARGB4444";
+  case OMX_COLOR_Format16bitARGB1555: return "OMX_COLOR_Format16bitARGB1555";
+  case OMX_COLOR_Format16bitRGB565: return "OMX_COLOR_Format16bitRGB565";
+  case OMX_COLOR_Format16bitBGR565: return "OMX_COLOR_Format16bitBGR565";
+  case OMX_COLOR_Format18bitRGB666: return "OMX_COLOR_Format18bitRGB666";
+  case OMX_COLOR_Format18bitARGB1665: return "OMX_COLOR_Format18bitARGB1665";
+  case OMX_COLOR_Format19bitARGB1666: return "OMX_COLOR_Format19bitARGB1666";
+  case OMX_COLOR_Format24bitRGB888: return "OMX_COLOR_Format24bitRGB888";
+  case OMX_COLOR_Format24bitBGR888: return "OMX_COLOR_Format24bitBGR888";
+  case OMX_COLOR_Format24bitARGB1887: return "OMX_COLOR_Format24bitARGB1887";
+  case OMX_COLOR_Format25bitARGB1888: return "OMX_COLOR_Format25bitARGB1888";
+  case OMX_COLOR_Format32bitBGRA8888: return "OMX_COLOR_Format32bitBGRA8888";
+  case OMX_COLOR_Format32bitARGB8888: return "OMX_COLOR_Format32bitARGB8888";
+  case OMX_COLOR_FormatYUV411Planar: return "OMX_COLOR_FormatYUV411Planar";
+  case OMX_COLOR_FormatYUV411PackedPlanar: return "OMX_COLOR_FormatYUV411PackedPlanar";
+  case OMX_COLOR_FormatYUV420Planar: return "OMX_COLOR_FormatYUV420Planar";
+  case OMX_COLOR_FormatYUV420PackedPlanar: return "OMX_COLOR_FormatYUV420PackedPlanar";
+  case OMX_COLOR_FormatYUV420SemiPlanar: return "OMX_COLOR_FormatYUV420SemiPlanar";
+  case OMX_COLOR_FormatYUV422Planar: return "OMX_COLOR_FormatYUV422Planar";
+  case OMX_COLOR_FormatYUV422PackedPlanar: return "OMX_COLOR_FormatYUV422PackedPlanar";
+  case OMX_COLOR_FormatYUV422SemiPlanar: return "OMX_COLOR_FormatYUV422SemiPlanar";
+  case OMX_COLOR_FormatYCbYCr: return "OMX_COLOR_FormatYCbYCr";
+  case OMX_COLOR_FormatYCrYCb: return "OMX_COLOR_FormatYCrYCb";
+  case OMX_COLOR_FormatCbYCrY: return "OMX_COLOR_FormatCbYCrY";
+  case OMX_COLOR_FormatCrYCbY: return "OMX_COLOR_FormatCrYCbY";
+  case OMX_COLOR_FormatYUV444Interleaved: return "OMX_COLOR_FormatYUV444Interleaved";
+  case OMX_COLOR_FormatRawBayer8bit: return "OMX_COLOR_FormatRawBayer8bit";
+  case OMX_COLOR_FormatRawBayer10bit: return "OMX_COLOR_FormatRawBayer10bit";
+  case OMX_COLOR_FormatRawBayer8bitcompressed: return "OMX_COLOR_FormatRawBayer8bitcompressed";
+  case OMX_COLOR_FormatL2: return "OMX_COLOR_FormatL2";
+  case OMX_COLOR_FormatL4: return "OMX_COLOR_FormatL4";
+  case OMX_COLOR_FormatL8: return "OMX_COLOR_FormatL8";
+  case OMX_COLOR_FormatL16: return "OMX_COLOR_FormatL16";
+  case OMX_COLOR_FormatL24: return "OMX_COLOR_FormatL24";
+  case OMX_COLOR_FormatL32: return "OMX_COLOR_FormatL32";
+  case OMX_COLOR_FormatYUV420PackedSemiPlanar: return "OMX_COLOR_FormatYUV420PackedSemiPlanar";
+  case OMX_COLOR_FormatYUV422PackedSemiPlanar: return "OMX_COLOR_FormatYUV422PackedSemiPlanar";
+  case OMX_COLOR_Format18BitBGR666: return "OMX_COLOR_Format18BitBGR666";
+  case OMX_COLOR_Format24BitARGB6666: return "OMX_COLOR_Format24BitARGB6666";
+  case OMX_COLOR_Format24BitABGR6666: return "OMX_COLOR_Format24BitABGR6666";
+  case OMX_COLOR_FormatKhronosExtensions: return "Khronos reserved value";
+  default: return "Unknown color";
+    }
+}
+
 static gboolean
 gst_omx_video_dec_negotiate (GstOMXVideoDec * self)
 {
@@ -1723,8 +1779,6 @@ gst_omx_video_dec_negotiate (GstOMXVideoDec * self)
   GstStructure *s;
   const gchar *format_str;
 
-  GST_DEBUG_OBJECT (self, "Trying to negotiate a video format with downstream");
-
   templ_caps = gst_pad_get_pad_template_caps (GST_VIDEO_DECODER_SRC_PAD (self));
   intersection =
       gst_pad_peer_query_caps (GST_VIDEO_DECODER_SRC_PAD (self), templ_caps);
@@ -1784,8 +1838,7 @@ gst_omx_video_dec_negotiate (GstOMXVideoDec * self)
     }
   }
 
-  GST_DEBUG_OBJECT (self, "Negotiating color format %s (%d)", format_str,
-      param.eColorFormat);
+  GST_DEBUG_OBJECT (self, "Negotiating color format %s (%s)\n", format_str, decode_color(param.eColorFormat));
 
   /* We must find something here */
   g_assert (l != NULL);
@@ -1799,6 +1852,10 @@ gst_omx_video_dec_negotiate (GstOMXVideoDec * self)
     GST_ERROR_OBJECT (self, "Failed to set video port format: %s (0x%08x)",
         gst_omx_error_to_string (err), err);
   }
+  else {
+    GST_DEBUG_OBJECT (self, "OMX_IndexParamVideoPortFormat %u : port %u, coding %#x, colorformat %s, framerate %#x \n",
+	   (unsigned int) param.nIndex, (unsigned int) param.nPortIndex, param.eCompressionFormat, decode_color(param.eColorFormat), (unsigned int) param.xFramerate );
+  }
 
   gst_caps_unref (intersection);
   return (err == OMX_ErrorNone);
-- 
2.9.0.137.gcf4c2cf

